### ISSUE TITLE
Ideas for maintaining parity with react-redux-starter-kit

### ISSUE THREAD

**Issue by davezuko** [SpencerCDixon/redux-cli#17]
First off, this is :sparkles: awesome :sparkles: , amazing work. Unfortunately I don't have the time right this second to type out as detailed of an issue as I'd like, but I wanted to put it out there in case you had any initial thoughts or ideas. Basically, what can be done in the base starter kit repo to make it easier to maintain feature consistency/parity?

For example, at work we have a fork of the starter kit and have built an updater utility that automatically migrates changes from the base template to the target project. Unfortunately it's not yet open sourced, so I'd have to talk to the team about what would have to be done to get something like that out the door and in good enough condition for public consumption.

I've been (_very_ slowly) working toward a more "stable" v2.0.0 release, but would be willing to tag a new version earlier if that makes sense for you, since, despite my best efforts, pulling latest master isn't always perfect. Basically, tell me what you need from me and I'll do my best to accommodate.

**Comment by SpencerCDixon** [comment]
Glad you like it!  Oh man I wish I had access to that updater!!! My repos that have used the starter kit are pretty far behind because I have been too lazy to try and keep it up to date/didn't know of a good way to go about it.  If there is any way your company would be willing to open source that I think that should be a priority.  It can act almost like a codemod in order for people to consistently upgrade their own apps with the kit & cli.  I'm totally open to help with that.  Having that would ease peoples minds in the whole 'which starter kit should I use' debate.  

I actually don't think it should be too hard to maintain consistency between the starter kit and the CLI.  Like you said, having the CLI pull specific versions of the starter kit is definitely the move.  A good next step might be for you to schedule a release and I can be working on the CLI to work well with that release.  I'm also down to wait until 2.0 comes out and I could bump the CLI to 2.0 and stay in sync from there if you think that makes the more sense.  

There are quite a few features I'd like to add to the CLI that arn't redux-starter-kit specific so even if it takes a few weeks to get to 2.0 that might even be ideal.

Philosophically it would be nice to establish some conventions.  I definitely don't think they need to be as intense as something like Rails but a big reason I started working on this is because I miss my Rails dev environment and how easy it was to walk into new projects/know the right thing to do.  Perhaps a good next step for that would be to start a 'conventions.md' doc or something where we can have the community throw out ideas/maybe even vote on conventions.

**Comment by ericwooley** [comment]
I saw on your roadmap you wanted to add the ability to overwrite templates anyways, that could be a good way to keep parity. 

Maybe add a .templates folder to react-redux-starter-kit and in that folder would be a dumb folder, a smart folder, redux-module folder etc...

The redux-cli could then look in the .templates folder of the project before grabbing it's own template. 

The workflow for keeping up to date as I do it is to just set the redux starter kit as upstream, pull and handle merge conflicts. This way the .template folder would get updated automatically and the redux cli would pick up on them without any kind of update.

This would also generalize the redux cli to not have to specifically follow react-redux-starter-kit, but any starter kit could follow this pattern.

**Comment by SpencerCDixon** [comment]
^ I like that idea a lot.  Thoughts @davezuko?  I was also thinking of changing `templates` to `blueprints` like in ember-cli.

**Comment by davezuko** [comment]
@SpencerCDixon sounds good to me. I will keep tracking this repo and try to get on it this weekend.

**Comment by SpencerCDixon** [comment]
:+1: doing a pretty big internal refactor of the cli after spending a few hours reading ember-cli

**Comment by ericwooley** [comment]
@SpencerCDixon Are you still doing a large refactor? I was thinking about adding some features, but don't want to conflict with a large refactor.

**Comment by SpencerCDixon** [comment]
Yeah!  Actually just merged it in recently.  Changed up how things work quite a bit to use blueprints.  Check out this PR into the react-redux-starter-kit: https://github.com/davezuko/react-redux-starter-kit/pull/620/files and here is some docs on how to create blueprints: https://github.com/SpencerCDixon/redux-cli#creating-blueprints 

Still need to finish test coverage on the blueprint model which is something I plan to do in the next day or two when I have some free time

**Comment by davezuko** [comment]
FYI @SpencerCDixon been another busy week for me, but for convenience just created a new tag https://github.com/davezuko/react-redux-starter-kit/releases/tag/v2.0.0-alpha.0 that you can target if you'd like. Will hopefully be able to get to improving some of the blueprints this weekend.

**Comment by SpencerCDixon** [comment]
Awesome sounds good, I will update the URI over the weekend.  Also we could totally add the like css files, etc. to the component/view blueprints :)

**Comment by nathanielks** [comment]
@SpencerCDixon @davezuko love the work y'all are doing! Was curious your thoughts on the blueprints before I got started on a PR: Forms, Views, and Layouts use the convention `layouts/{name}Layout/{name}Layout.js`. How do y'all feel if it were `layouts/{name}Layout/index.js`? This way all you'd need to do is this:

```js
import CoreLayout from 'layouts/CoreLayout'
```

Instead of:

```js
import CoreLayout from 'layouts/CoreLayout/CoreLayout'
```

An added benefit is each of the [insert thing here] could also define specific exports if they needed to?

**Comment by davezuko** [comment]
This comes up a lot, and is -- to my understanding -- currently possible with the current blueprint design, so you would just have to customize it to your own preference. There are some shortfalls with it, however:

* If users have a lot of files open in their IDE, the `index.js` pattern means they'll just have a bunch of `index.js` tabs open, which is difficult to parse. Some editors are smart enough to prefix these files with the path, but not all are.
* Similarly, this can make searching for the file you want more difficult.

I used to fall under your camp as well, as it does make imports _much_ friendlier, but after working on a team that explicitly names each file I've come to appreciate some of the benefits. So, all that said, that's just _my_ preference, and it all just comes back to preference. Not sure what else to add here, other than that if it does get switched, there will inevitably be a PR to switch back to the other direction, speaking from experience ;).

**Comment by nathanielks** [comment]
@davezuko no, that totally makes sense! I hadn't thought of the open-tabs-readability! DX makes a lot of sense here :+1:

**Comment by nathanielks** [comment]
@davezuko One more potential optimization that satisfies both: `layouts/{name}/{name}Layout.js`?

**Comment by nathanielks** [comment]
Seeing as `resolve.root` is enabled, I _believe_ the import will always have the context of what you're importing, eg layouts, views, forms, etc. The filename will also be more indicative of what you're working on, so that way your tabs are readable too!

**Comment by davezuko** [comment]
That also works. Maybe the best solution is just to provide some out-of-the-box convenience for switching among the four patterns without rewriting all of the blueprints:

* Not nested (`layouts/CoreLayout.js`)
* Fully qualified (`layouts/CoreLayout/CoreLayout.js`)
* Node-style (`layouts/CoreLayout/index.js`)
* Qualified (shortened): (`layouts/Core/CoreLayout.js`)

Though it's kind of a one-and-done sort of thing, so it might be too much to include for not much gain, since there's bound to be even more patterns that people want to use.

**Comment by davezuko** [comment]
Maybe a poll on most popular convention to use as default?

**Comment by nathanielks** [comment]
that sounds like a great call :+1:

**Comment by SpencerCDixon** [comment]
Yeah a poll to decide would be great as far as the base blueprints in the starter kit go!  @nathanielks like dave said, the blueprints are designed in a way to be easy to customize for your teams specific needs if you just want it changed for a specific project.  Seems like an uphill battle to try and please everyone and the React community has so much disparity over conventions it drives me nuts...

Initially I really wanted to make the CLI have strict conventions in order to help the ecosystem build tooling around a specific set of conventions so I can get back to my 'Rails speed' of development.  Not really sure if or when that will happen though just due to the nature of React.

**Comment by nathanielks** [comment]
@SpencerCDixon I'm _right_ there with you. It's absolutely killing me we still haven't arrived at a common set of patterns. To add to the insanity, I'm actually using this alongside https://github.com/bdefore/universal-redux because I like that the webpack/server configuration is contained in a module, but I also like how redux-cli/react-redux-starter-kit lints and organizes everything!

**Comment by nathanielks** [comment]
@davezuko Aha! I just realized we could have the specific filename as well as use an index.js!

`layouts/{name}/{name}Layout.js`
`layouts/{name}/index.js`:
```js
import {name} from './{name}.js'
export default {name}
```

**Comment by SpencerCDixon** [comment]
:+1: I'm all for it!

**Comment by nathanielks** [comment]
Brilliant! @davezuko does that sound good to you?

**Comment by davezuko** [comment]
:+1:

**Comment by nathanielks** [comment]
I'll put a PR together!

**Comment by nathanielks** [comment]
Just realized the blueprints here aren't the same as in react-redux-starter-kit. @SpencerCDixon do you want those in here as well?

**Comment by SpencerCDixon** [comment]
Hmmm, I think I would leave the blueprints as they stand in the redux-cli repo since they're just meant to be super generic and use the smart/dumb paths for people to see how blueprints work.

Probably makes most sense to submit PR on the starter kit to help the kit have a robust set of blueprints thus making it more useful to more people

**Comment by dbroadhurst** [comment]
With the removal of Blueprints in react-redux-starter-kit I'm getting an error with

```
redux g dump Hello
throw new Error('Unknown blueprint: ' + name);
```

How should I use redux-cli with  react-redux-starter-kit?

**Comment by SpencerCDixon** [comment]
You don't need to use it with that project if you don't want.  If you'd like to create your own blueprints for your own personal project (which is what I do) then by all means go for it!

**Comment by anithri** [comment]
In 2.0, I want to be able to read from multiple blueprint directories, especially including npm modules.  This would allow for sharing sets of blueprints and make it easier for users to select which style of blueprints they wish to use.  This would allow us to create sets of BP for each of the `new` templates customized.

We should add a set matching (create-react-app)[https://github.com/facebookincubator/create-react-app].

**Comment by anithri** [comment]
Blueprint packages are being added for 2.0.  This will allow groups of blueprints to be shared and distributed with npm. Groups of blueprints that closely adhere to a given standard, like the start kit, should be packaged together.  

We'll start documentation tracking all the blueprint packages we know about.

Discussion what blueprints should be included by default in #114, I'm closing this issue in preference to that one.

**Comment by wjljack** [comment]
@anithri I use lastest 2.0 repo ,and use `npm link` to link to global,I find that in version 1.9.0,`redux g xxx name` command find template in npm/node_modules/redux-cli/blueprints,but in 2.0 ,`blueprint g xxx name` can only find template files in current blueprints directory,not in global node_modules directory,when i use `blueprint g xxx name` ,it show `Unknown blueprints xxx`,xxx is in global node_modules/redux-cli/blueprints,not in current project directory,how can i use 2.0 to read both global node_modules/redux-cli template and current project /project_root/blueprints template?
